Implement configurations and portable mode. Closes #465

evsh opened the pull request:
It may be useful to have different configurations either for portable versions or for debugging purposes. To implement this we add two options, avaliable via command line switches
1. An option to change configuration name (`--configuration`). The name supplied via this option is appended to `QCoreApplication::applicationName()` to form `qBittorrent_<conf_name>` name for the configuration files.
2. An option to provide a path do directory where all the settings are stored (kind of profile directory). There is a shortcut `--portable` which means "use directory `profile` near the executable location".

In order to implement that we have to perform initialisation of the profile directories before the `SettingStorage` and `Preferences` singletones are initialised. Thus, options parsing shall be performed without defaults read from preferences.

Update:
Added command line options to implement both the different configurations and portable mode.

evsh commented on src/base/utils/fs.cpp:
```diff
@@ -518,3 +519,9 @@ QString Utils::Fs::cacheLocation()
+
+QString Utils::Fs::configurationName()
+{
+    QString envConfName = QString::fromLocal8Bit(qgetenv("QBT_CONF_NAME"));
```
It might make sense to make this variable static. However, I can't imagine qBt changing this environment variable at run time.

glassez replied:
Change this in runtime will cause undefined behaviour or even crash the program. So it must be initialized at program startup and do not change further.

sledgehammer999 commented:
Unless I am mistaken you actually implement #465 in some way. However, under this light an environment variable isn't very desirable :p

evsh commented:
Hm... Kind of... I just tired of qBt rechecking my files over and over again. 

As to me, the environment variable approach gives flexibility for developers. One can have separate configurations for different tests, check how qBt behaves with an empty configuration (by the way, it crashes with LC_COLLATE=C when opening the options dialogue) in no time.

Do you suggest to add a command line switch to set this variable for the portable version?

evsh commented:
> Do you suggest to add a command line switch to set this variable for the portable version?

And, of course, a code that will check that variable contents for an absolute path, which will be used as is but not appended.

sledgehammer999 commented:
>check how qBt behaves with an empty configuration (by the way, it crashes with LC_COLLATE=C when opening the options dialogue) in no time.

This shouldn't happen. Clean installs should just work.

>Do you suggest to add a command line switch to set this variable for the portable version?

If it isn't trouble for you. Am I wrong or did the last merges greatly simplified the code for "portable version" implementation?

>And, of course, a code that will check that variable contents for an absolute path, which will be used as is but not appended.

In portable versions the config could reside in the same folder as the binary. So relative paths should work too.

evsh commented:
Configuration name is appended to the paths like `QDir::homePath()`.

evsh commented:
I'm going to pack  Utils::Fs::QDesktopServices implementations into a class, add an override version with manually specified root path and instantiate one or another basing on the command line switch. What do you think?

evsh commented:
@qbittorrent/qbittorrent-frequent-contributors Can someone do me a favor, please, and make extensive tests on Windows and OS X for the portable version?

glassez commented on src/base/utils/fs.h:
```diff
@@ -67,6 +67,10 @@ namespace Utils
+
+        /// Returns either default name for configuration file (QCoreApplication::applicationName())
+        /// or the value, supplied via QBT_CONF_NAME environment variable
+        QString configurationName();
```
This function, like some other (cacheLocation, QDesktopServices\*) , needs to be moved to another place, because they sense are not file system utilities. These are shell or environment utilities.

glassez commented on src/base/settingsstorage.cpp:
```diff
@@ -70,9 +70,9 @@ namespace
         {
 #ifdef Q_OS_WIN
-            return SettingsPtr(new QSettings(QSettings::IniFormat, QSettings::UserScope, "qBittorrent", name));
+            return SettingsPtr(new QSettings(QSettings::IniFormat, QSettings::UserScope, Utils::Fs::configurationName(), name));
```
SettingsStorage is not the only class that requires modifications. There are some other classes that store all their settings/state/data in the same place. Although it may be enough changes in QDesktopServicesDataLocation for them.

evsh commented:
PR updated. Now it implements portable mode. Needs testing. 

@sledgehammer999: support for command line switches has been added, but environment variables are supported too. I believe they are better for debugging and may be of use for users too.

Two settings are proposed: 
`--profile=<dir>` allows to set configuration prefix. Inside of it directories `cache`, `config`, `data`, `downloads` are created. If the switch is not supplied, default system dirs are used.
Option `--portable` is a shortcut for `--profile=<exe dir>/profile`. 
--configuration=<name> allows to have different configuration inside of a configuration prefix.

If `--profile` option is not empty, `SettingStorage` serialises settings in `QSettings::IniFormat` to make them portable between platforms.

evsh commented:
There is a destructive change: usage help text does not contain the Web UI port number from settings anymore because options parsing needs to be done before preferences initialisation.

glassez commented on src/base/utils/environment.h:
```diff
@@ -0,0 +1,88 @@
+        namespace impl
+        {
+            class SpecialFoldersImpl;
+            class SpecialFoldersSingletongImpl;
```
Singleton**g**?

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -377,137 +360,17 @@ QString Utils::Fs::expandPathAbs(const QString& path)
 
 QString Utils::Fs::QDesktopServicesDataLocation()
```
Why do you leave these ugly wrappers?
Why shouldn't we use Environment utils directly?

evsh replied:
OK.

glassez commented on src/base/utils/environment.cpp:
```diff
@@ -0,0 +1,79 @@
+
+Utils::Envrironment::SpecialFolders*Utils::Envrironment::SpecialFolders::m_instance = nullptr;
+
+Utils::Envrironment::SpecialFolders::~SpecialFolders()
```
using namespace Utils::Environment is good idea!

evsh commented on src/base/utils/fs.cpp:
```diff
@@ -375,144 +358,9 @@ QString Utils::Fs::expandPathAbs(const QString& path)
-#endif
-}
-
 QString Utils::Fs::cacheLocation()
```
I don't understand this function and thus left it as is.

glassez replied:
It returns the cache dir path and create this directory if it doesn't exist which is overkill. So you may drop this function and use your Environment utils directly.

evsh replied:
But why it was written? Apparently, somebody encountered a situation when the cache directory was missing. Should we create them all the directories when we create the `SpecialFoldersImpl` object?

evsh replied:
> Should we create them all the directories when we create the SpecialFoldersImpl object?

After some thinking I like this approach. In `Utils::Envrironment::SpecialFolders::initialize()` we ensure that all paths exist in the same way as we do in `Utils::Envrironment::impl::SpecialFoldersImplCustom::SpecialFoldersImplCustom` and remove all the `QDir::mkpath()` calls from the individual functions. The program can not run without the config, cache, and data directories. As for the downloads directory, we already create it now anyway.

evsh commented:
This closes #4887 too

evsh commented:
@qbittorrent/qbittorrent-frequent-contributors:
This is more or less ready. With two exceptions (see below) `strace qbittorrent --profile /tmp/qbt_profile` does not show any app-specific file activity in ${HOME}.

The first exception is `${HOME}/.config/qBittorrentrc`, but I guess it is used by KF5, and we can do nothing about it. The file keeps `KFileDialog` settings.

The second exception is the `QIniSettings` class. It is used in statistics and RSS. And my question is: may I remove the `QIniSettings` class? Why is it needed? May I replace it with `Utils::Environment::SpecialFolders::applicationSettings()` calls? We can add an overloaded `applicationSettings()` for that cases to pass `QSettings::Format`.

Chocobo1 commented on src/app/options.h:
```diff
@@ -0,0 +1,70 @@
+ *
+ * Contact : chris@qbittorrent.org
+ */
+
```
Is include guard omitted on purpose?

evsh replied:
No, it was a mistake. Thanks!

Chocobo1 commented on src/base/utils/environment.h:
```diff
@@ -0,0 +1,87 @@
+
+            /// Returns either default name for configuration file (QCoreApplication::applicationName())
+            /// or the value, supplied via parameters
+            QString configurationName();
```
You can add `const` here.

Chocobo1 commented on src/base/utils/environment_p.cpp:
```diff
@@ -0,0 +1,292 @@
+Utils::Envrironment::SettingsPtr Utils::Envrironment::impl::SpecialFoldersImplCustom::applicationSettings(const QString &name) const
+{
+    // here we force QSettings::IniFormat format because we need it to be portable across platforms
+    return SettingsPtr(new QSettings(QDir(configLocation()).absoluteFilePath(name + QLatin1String(".conf")), QSettings::IniFormat));
```
maybe just me, you could use a temporary QString to hold the QDir::absoluteFilePath(), will be easier to read.

Chocobo1 commented on src/base/utils/environment_p.h:
```diff
@@ -0,0 +1,98 @@
+                QDir m_rootDirectory;
+                static const QString cacheDirName;
+                static const QString configDirName;
+                static const QString dataDirName;
```
```c++
const QString dataDirName = QLatin1String("data");  // Why not this?
```

Chocobo1 commented on src/app/main.cpp:
```diff
@@ -90,236 +92,139 @@ const char *sysSigName[] = {
-        }
+        return app->exec(params.torrents);
+    }
+    catch (CommandLineParameterError &er) {
```
This `catch` is quite away from the `throw`s, is it a good idea to move up?

evsh replied:
Agree, the body of `main()` is too long, we better split it onto several functions.

Chocobo1 commented on src/app/options.h:
```diff
@@ -0,0 +1,70 @@
+    QBtCommandLineParameters();
+};
+
+class CommandLineParameterError: public std::runtime_error
```
can you share the reason of handling error by exceptions? I'm curious about it.

evsh replied:
Two reasons: I had to signal error from `Application::Application()` and `main()` had many repeative calls to `displayBadArgMessage()` with `return EXIT_FAILURE` (that is in `catch` clause now).

Chocobo1 commented on src/base/utils/environment.h:
```diff
@@ -0,0 +1,87 @@
+        class SpecialFolders
+        {
+        public:
+            QString location(SpecialFolder folder) const;
```
`const SpecialFolder folder`?

evsh replied:
Sorry, can not understand why it would be better with `const`.

Chocobo1 replied:
By reading the method signature I would suspect `folder` will somehow change to other value, but it didn't.
Well this is a habit of mine...

glassez replied:
>By reading the method signature I would suspect folder will somehow change to other value, but it didn't.

You read the method signature incorrectly! The `folder` cannot be changed since it passed by value. The method free to change its `folder` copy.

Chocobo1 replied:
>You read the method signature incorrectly! 

You must have misunderstood me.
I meant I suspect the local copy `folder` will somehow change to other value but it didn't.

glassez replied:
Passing param by value as const creates hard dependency between function signature and its implementation. If in the future we will need to do some preprocessing of this param, we will either have to change the method signature or use some tricks (create additional variables, etc.).

evsh replied:
I hope that even with this polemic you guys will find a time to look at `QIniSettings` and think whether we can drop it completely, please? :)

Chocobo1 replied:
@glassez 
That sounds like you don't like to add `const` for pass-by-value parameters!?

glassez replied:
>That sounds like you don't like to add const for pass-by-value parameters!?

Yes. Except some really special cases.

glassez commented:
>This is more or less ready.

@evsh less than you think.
Your biggest error here is that you changed the order of things in `main()`, without knowing how it actually works.
Application instance MUST be created BEFORE we parse command line since QApplication should parse and filter the parameters destined for it. Otherwise you will get an exception if the user wants to specify such command line parameters. The second (minor) mistake is you cannot use any translations before Application is initialized.
I'm not going to dwell on other (less important) issues. First you need to fix this.

evsh commented:
> Application instance MUST be created BEFORE we parse command line since QApplication should parse and filter the parameters destined for it. Otherwise you will get an exception if the user wants to specify such command line parameters.

Can't understand you, sorry. Could you explain it, please?

glassez commented:
>The second exception is the `QIniSettings` class. It is used in statistics and RSS. And my question is: may I remove the `QIniSettings` class? Why is it needed?

If you look carefully the code, it becomes clear that it is completely useless. That's why I refused to use it in `SettingsStorage` class.
>May I replace it with Utils::Environment::SpecialFolders::applicationSettings() calls?

IIRC, yes.

BTW, the last time we use the `QSettings` class not for its intended purpose, and only for serialization of settings into the file. It's all the same to shoot with a Bazooka on the wheel. Isn't it time for us to think about using something more suitable? @sledgehammer999, what do you say?

glassez commented:
>Can't understand you, sorry. Could you explain it, please?

For example, QApplication recognizes `-style=<style>` argument. IIRC, If user pass it your code throws an exception since it unknown option for qBittorrent. But if you bypass command line to QApplication constructor it filters all known options and modifies `argc`/`argv`. Then you can do custom parsing the rest.

evsh commented:
Oh, I get it now. Thanks! Indeed, I forgot about tanslations and Qt options completetly.

evsh commented:
> If you look carefully the code, it becomes clear that it is completely useless.

I did look, did make the same conclusion and that is why I'm asking.

glassez commented:
>Oh, I get it now. Thanks! Indeed, I forgot about tanslations and Qt options completetly.

You should keep current order and try other solution. E.g. you may parse command line in Application constructor and then use parsed params via some Application method (something like `Application::commandLineArgs()`).

evsh commented:
Yes, thanks. It is obvious that parameters parsing has to be done in `Application::Application()`, because this is the onle place between `QApplication::QApplication()` and `Preferences` initialisation.

glassez commented:
>Option --portable is a shortcut for --profile=\<exe dir\>/profile.
--configuration=\<name\> option allows to have different configuration inside of a configuration prefix.

What happens if the user set `--portable` and `--profile=some_path` together?

evsh commented:
`Application::Application()` [throws](https://github.com/qbittorrent/qBittorrent/pull/5214/files#diff-83197ed34f0ea23a0ccd7edfa47822eeR103) an exception which is shown to the user and the program aborts.

glassez commented:
>Application::Application() throws an exception which is shown to the user and the program aborts.

Ok.

glassez commented:
@evsh For some reason we need to know the original number of parameters. This can only be done before calling the QApplication constructor.

evsh commented:
>  For some reason we need to know the original number of parameters.

Actually, I don't understand this. If a user wants to display the dialogue with the usage text using specific style or stylesheet, why not?

evsh commented:
PR updated: options parsing was moved into the `Application` constructor, `QIniSettings` class was removed.

evsh commented:
And I would like to add that all the playing with characters in `QString`s after working with `boost::filesystem::path` is like using C after C++.

glassez commented:
>Actually, I don't understand this. If a user wants to display the dialogue with the usage text using specific style or stylesheet, why not?

Ok.

glassez commented:
@evsh, the user downloads the torrent to removable disk beside portable qBittorrent, and he continues to do this on another computer - it's real use case of portable app?
If yes your implementation is incomplete since we store absolute save paths in resume data.

glassez commented:
>And I would like to add that all the playing with characters in  QString s after working with  boost::filesystem::path  is like using C after C++.

Unfortunately it's incompatible with QString and we would often convert strings to use it. We also can implement QString/QDir based Path class.

evsh commented:
> Unfortunately it's incompatible with QString and we would often convert strings to use it. We also can implement QString/QDir based Path class.

The change would make sense only together with replacing the whole Qt I/O, of course.

evsh commented:
> @evsh, the user downloads the torrent to removable disk beside portable qBittorrent, and he continues to do this on another computer - it's real use case of portable app?
If yes your implementation is incomplete since we store absolute save paths in resume data.

Do you mean libtorrent and its "save_path" attribute?

glassez commented:
>Do you mean libtorrent and its "save_path" attribute?

Both libtorrent save_path and qBittorrent qBt-savePath.

evsh commented:
Would it be OK if I hook up into the save/load sequences and replace both attributes with a path, relative to the $HOME or profile dir?

evsh commented:
> Would it be OK if I hook up into the save/load sequences and replace both attributes with a path, relative to the $HOME or profile dir?

Oh, no! Windows might have several filesystem roots! What to do then?

glassez commented:
>The change would make sense only together with replacing the whole Qt I/O, of course.

Well, in that case, I prefer not to continue further on this topic.
>Oh, no! Windows might have several filesystem roots! What to do then?

We need to carefully consider it. I have no solution yet...

glassez commented:
@evsh, you can abandon claims for portable mode here, complete everything else, and then you can start to think about it.

evsh commented:
> We need to carefully consider it. I have no solution yet...

I see it in the following way. On saving:
`<windows only>` If `profile_dir` drive letter is the same as save_path drive letter`</windows only>` make save_path relative to the `profile_dir`.
On loading:
If save_path is relative, make it absolute using profile dir.

`profile_dir` is `${HOME}` or the one specified via `--profile`/`${QBT_PROFILE}`

evsh commented:
The alternative is to make path relative only if it is inside of `profile_dir`

evsh commented:
> @evsh, you can abandon claims for portable mode here, complete everything else, and then you can start to think about it.

If we are one step from it, why dont't we make it work?

Chocobo1 commented:
>@evsh, the user downloads the torrent to removable disk beside portable qBittorrent, and he continues to do this on another computer - it's real use case of portable app?

Yes. 
This usage is interesting but I think its optional not mandatory.

IMO, if an app can be moved to anywhere without losing its settings, it can be called a portable app. This doesn't imply the artifacts that the app produces is also portable.

Here's another scenario I can think of: qbt app was in dir `C:\qbt` and seed torrents in `e:\`, now the user move app dir to `d:\qbt`, then the user might expect the seeding continue to work.

glassez commented:
>The alternative is to make path relative only if it is inside of profile_dir

It is a good idea.
Then we can just set `profile_dir` as working directory when the application is run, if `--portable` is set.

evsh commented:
But that rules out interesting cases with qbt on an external drive, where it is not convenient to download inside of the qbt directory. I incline towards the first option (relative path if on the same drive).

glassez commented:
> I incline towards the first option (relative path if on the same drive).

But only if `--portable` parameter is set.
But we could offer the user to make a choice when adding a torrent (in portable mode).

glassez commented:
@evsh when you're done most of the work here, check and fix coding style. The main mistakes that I noticed a cursory glance, is naming (namespace names start with Capital letter) and statement breaks (operators are placed not at the end of the line and at the beginning of the next one).

evsh commented:
> But only if --portable parameter is set.

@glassez why? what shortcomings do you see?

glassez commented:
>@glassez why? what shortcomings do you see?

I just don't want qBittorrent was doing some extra work in the normal (non-portable) mode.

evsh commented:
We can add two methods
```C++
QString toPortablePath(const QString &absolutePath);
QString fromPortablePath(const QString &portablePath);
```
to the `Profile` class (see the updated diffs). Those in the `Impl::DefaultProfile` will do nothing.

glassez commented:
They should do nothing even if we set `--profile` but not `--portable`.

evsh commented:
Disagree with you: current meanings of `--portable` and `--profile` are identical.

glassez commented:
>current meanings of --portable and --profile are identical.

Really?
>Option --portable is a shortcut for --profile=\<exe dir\>/profile

They are identical in only one case. And the rest?
When I want to use a different profile (e.g. for testing), it does not mean that I want to enable portable mode.

evsh commented:
> Really?

Yes, they both set profile directory.
 
> They are identical in only one case. And the rest?
When I want to use a different profile (e.g. for testing), it does not mean that I want to enable portable mode.

For testing you may use different configurations. I'd better add an options `--no-portable-fastresume`, because otherwise most of the users of portable mode would need to pass two options. While a developer could use the second command line argument/environment variable.

evsh commented:
Or, we add `--portable-fastresume` and `--portable` implies `--profile=x` *and* `--portable-fastresume`.

glassez commented on src/app/options.h:
```diff
@@ -28,46 +30,47 @@
+    QString unknownParameter;
 
-  }
+    QBtCommandLineParameters(const QProcessEnvironment&);
```
Maybe remove `QProcessEnvironment` from constructor params and us `QProcessEnvironment::systemEnvironment()` in constructor body? You don't like yourself the extra things in the interface!

evsh replied:
I wanted to point out that default values are loaded from environment.

glassez commented on src/app/options.h:
```diff
@@ -28,46 +30,47 @@
+{
+public:
+    CommandLineParameterError(const QString &messageForUser);
+    const QString& messageForUser() const;
```
There are still some messages? Why such verbosity? Why not just `message`?

evsh replied:
Usually, exception texts are for developers, that is why.

glassez commented:
>Yes, they both set profile directory.

But only one set portable mode!
>For testing you may use different configurations. I'd better add an option --no-portable-fastresume, because otherwise most of the users of portable mode would need to pass two options. While a developer could use the second command line argument/environment variable. 
Or, we add --portable-fastresume and --portable implies --profile=x and --portable-fastresume.

What is all this? Why you need to have a separate `--portable-fastresume`?
>otherwise most of the users of portable mode would need to pass two options

Why? They should pass only `--portable`. Other (normal mode users) may pass `--profile=<profile_path>`. And that's all!

evsh commented:
> Why? They should pass only --portable.

Because portable mode user may want to keep the qBt profile not in the default location but somewhere else.

glassez commented:
>Because portable mode user may want to keep the qBt profile not in the default location but somewhere else. 

Ok. Then you just should allow to set `--profile' together with `--portable`.
But, IMO, if you go on about all the desires of the user, nothing good happens.

evsh commented:
A few moments ago you've pointed out that profile directory and relative paths in the fastresume files are independent options. Why don't we associate each with a command line parameter then? That's all. No more variations, no more desires. That would be logical, IMO. And then we introduce the `--portable` switch to give reasonable defaults for portable configurations.

glassez commented:
And then we introduce lot of possible undefined behaviors!

>A few moments ago you've pointed out that profile directory and relative paths in the fastresume files are independent options. 

Yes. But "relative paths in the fastresume files" is not separate option, it's just attribute of "portable mode" option.

evsh commented:
> And then we introduce lot of possible undefined behaviors!

Namely?

> Yes. But "relative paths in the fastresume files" is not separate option, it's just attribute of "portable mode" option.

I begin to suspect that the name `--portable` for the command line switch is wrong :)  "relative paths in the profile directory" is the attribute of thr portable mode as well.

glassez commented on src/app/application.cpp:
```diff
@@ -97,19 +98,33 @@ Application::Application(const QString &id, int &argc, char **argv)
+
+    setApplicationName("qBittorrent");
+
+    QString profileDir = m_commandLineArgs.portableMode ?
```
Fix coding style:
```c++
QString profileDir = m_commandLineArgs.portableMode
            ? QDir(QCoreApplication::applicationDirPath()).absoluteFilePath(QLatin1String("profile"))
            : m_commandLineArgs.profileDir;
```

evsh replied:
Are  you thinking that '?' on the next line adds readability?

glassez replied:
I am thinking we should following the coding guidelines.
I am thinking the code uniformity adds readability.

glassez commented on src/app/application.cpp:
```diff
@@ -153,7 +173,7 @@ void Application::setFileLoggerEnabled(bool value)
 QString Application::fileLoggerPath() const
 {
-    return settings()->loadValue(KEY_FILELOGGER_PATH, QVariant(Utils::Fs::QDesktopServicesDataLocation() + LOG_FOLDER)).toString();
+    return settings()->loadValue(KEY_FILELOGGER_PATH, QVariant(specialFolderLocation(Utils::Environment::SpecialFolder::Data) + LOG_FOLDER)).toString();
```
It's better to split such long lines.

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,394 @@
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2014  Vladimir Golovnev <glassez@yandex.ru>
+ * Copyright (C) 2006  Christophe Dumez
+ *               2016  Eugene Shalygin
```
Why do you love to violate the order? The copyright notice should be placed in reverse chronological order and be as follows:
```
Copyright (C) 2016  Eugene Shalygin <your_email>
```
You can see how it's done!

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,394 @@
+
+        QString envVarName() const
+        {
+            return QLatin1String("QBT_") +
```
Fix coding style!

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    // Option with string value. May not have a shortcut
+    struct StringOption: protected Option
```
Why `protected`?

evsh replied:
`BoolOption` is an exception (with  its public inheritance). But probably I just need to move `Option::usage()` from `Option` to `BoolOption`.

evsh replied:
Done.

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,394 @@
+
+        static bool isStringQuoted(const QString &s, QChar quoteChar)
+        {
+            return s.size() >= 2 && (s.startsWith(quoteChar) && s.endsWith(quoteChar));
```
Please add parenthesis around `s.size() >= 2` too.

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,394 @@
+    for (int i = 1; i < args.count(); ++i) {
+        const QString &arg = args[i];
+
+        if ((arg.startsWith("--") && !arg.endsWith(".torrent")) ||
```
Fix coding style!

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,394 @@
+#endif
+            else if (arg == WEBUI_PORT_OPTION) {
+                result.webUiPort = WEBUI_PORT_OPTION.value(arg);
+                if (( result.webUiPort < 1) || ( result.webUiPort > 65535) )
```
Fix coding style!

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,394 @@
+    return m_messageForUser;
+}
+
+QString makeUsage(const QString &prg_name)
```
Please fix param name (use camelCase).
I understand that most likely it got you inherited, but I hope it's not too much trouble you.

glassez commented on src/app/options.h:
```diff
@@ -28,46 +30,47 @@
-  QIniSettings(const QString &organization = "qBittorrent", const QString &application = "qBittorrent", QObject *parent = 0 ):
-#ifdef Q_OS_WIN
-      QSettings(QSettings::IniFormat, QSettings::UserScope, organization, application, parent)
+class QProcessEnvironment;
```
Insert blank line after.

evsh replied:
Done.

glassez commented on src/base/utils/environment.cpp:
```diff
@@ -0,0 +1,103 @@
+
+#include "environment_p.h"
+
+Utils::Environment::Profile*Utils::Environment::Profile::m_instance = nullptr;
```
Add space before `*`.

glassez commented on src/base/utils/environment.cpp:
```diff
@@ -0,0 +1,103 @@
+
+Utils::Environment::Profile*Utils::Environment::Profile::m_instance = nullptr;
+
+Utils::Environment::Profile::Profile(Impl::Profile *impl)
```
This file is an implementation of the classes in the `Utils::Environment` namespace. To use `using namespace Utils::Environment` is a good idea here.

glassez commented on src/base/utils/environment.h:
```diff
@@ -0,0 +1,94 @@
+            Profile(Impl::Profile *impl);
+            ~Profile();
+
+            friend class ::Application;
```
Base MUST be independent from App, GUI etc.

evsh replied:
Who then can initialize this singleton, if not Application?

glassez replied:
Application initializes (almost) all current singletons, but they are independent from it. Just do `initialize()` public.

evsh replied:
Bad idea. IMO, a friend declaration, in this case, is not a dependence. Consider the following example:
```C++
class A {
private:
   friend int main(int, char**);
   static void initialize();
};
```
Does `A` depend on `main()`? No, it utilises the idea that there *has* to be the `main()` function and only it can initialise the class. The same is here: everyone knows that there has to be an `Application` class in Qt application.

glassez replied:
Okay, probably you're right here...

evsh replied:
Glad that you agree.

glassez commented on src/base/utils/environment_p.cpp:
```diff
@@ -0,0 +1,282 @@
+}
+
+Utils::Environment::SettingsPtr
+Utils::Environment::Impl::DefaultProfile::applicationSettings(const QString &name) const
```
As I mentioned above, better to remove `Utils::Environment` and `Utils::Environment::Impl` from these names, than so to be perverted here.

glassez commented on src/base/utils/environment_p.cpp:
```diff
@@ -0,0 +1,282 @@
+
+Utils::Environment::Impl::CustomProfile::CustomProfile(const QString &rootPath, const QString &configurationName)
+    : Profile(configurationName)
+    , m_rootDirectory {QDir(rootPath).absoluteFilePath(this->configurationName())}
```
Please don't mix initializer styles.

glassez commented on src/base/utils/environment_p.h:
```diff
@@ -0,0 +1,101 @@
+                QString configurationName() const;
+
+            protected:
+                Profile(const QString &configurationName);
```
As **Profile** is abstract class you may have public constructor here.

evsh replied:
I might misunderstand you, but there is no public constructor because this is an abstract class.

glassez replied:
I mean you can have public constructor in abstract class. No need extra (protected) section here.

evsh replied:
You probably mean that it is impossible to create and instance with pure virtual functions anyway. IMO,  protected constructor in such case is better for documentation/hint reasons.

glassez replied:
>IMO, protected constructor in such case is better for documentation/hint reasons. 

Okay, if so.

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -51,23 +47,10 @@
-
-#include "misc.h"
-#include "fs.h"
+#include "environment.h"
```
This isn't needed.

evsh commented on src/base/bittorrent/session.cpp:
```diff
@@ -2686,7 +2687,8 @@ namespace
         if (ec || (fast.type() != libt::lazy_entry::dict_t)) return false;
 
-        torrentData.savePath = Utils::Fs::fromNativePath(Utils::String::fromStdString(fast.dict_find_string_value("qBt-savePath")));
+        torrentData.savePath = Utils::Environment::Profile::instance().fromPortablePath(
```
@glassez: could you check these changes, please?

evsh commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -1531,7 +1532,13 @@ void TorrentHandle::handleSaveResumeDataAlert(libtorrent::save_resume_data_alert
         resumeData["qBt-forced"] = isForced();
     }
-    resumeData["qBt-savePath"] = m_useASM ? "" : Utils::String::toStdString(m_savePath);
+    else {
```
@glassez: and here too, please.

Chocobo1 commented on src/base/utils/environment_p.cpp:
```diff
@@ -280,3 +280,44 @@ Utils::Environment::Impl::CustomProfile::applicationSettings(const QString &name
+    if (QDir::isAbsolutePath(path)) {
+        QChar driveLeter = path[0].toUpper();
+        QChar baseDriveLetter = m_baseDir.path()[0].toUpper();
+        bool onSameDrive = (driveLeter.category() == QChar::Letter_Uppercase) && (driveLeter == baseDriveLetter);
```
I wasn't going to question it anymore, but I can't hold myself :P
Please, just for once; could you explain (convince) me that this implementation will work fine to my scenario in: https://github.com/qbittorrent/qBittorrent/pull/5214#issuecomment-218716346

I believe that is a very common usage scenario.
Or let me know why my scenario might be invalid.

evsh replied:
I guess you imply `--portable` switch set. Things will not work in this scenario (obviously). That is why we split the portable mode onto two options.

Chocobo1 replied:
>That is why we split the portable mode onto two options.

I am confused, how do `--configuration` & `--profile` help for this scenario?

evsh replied:
There are four options added by this PR. Two of them turn on independent features of the portable mode: `--profile` and `--relative-fastresume`. There is one option for configurations (`--configuration) and there is a shortcut `--profile` which is equal to `----profile=$EXE_DIR/profile` and `--elative-fastresume`.

For the use-case noted by @glassez one needs just `--profile` option. Those who use a portable drive for qbt and downloads need both options.

Chocobo1 commented:
Compiled & tested, the command line switches works fine!

Some minor document issues:
1. in command line help, the Options list should have sorted alphabetically.
2. `For example, to disable the splash screen: QBT_NO_SPLASH=1 qbittorrent`, the last `qbittorrent` should be replaced by `argv[0]`.
3. `Bad command line: Portable mode and explicit profile directory options are mutually exclusive`, the incompatibility of 2 options should have been mentioned under `--portable` or `--profile` point.
4. `Bad command line: Portable mode and explicit profile directory options are mutually exclusive`, here the `Portable mode` & `explicit profile` are mentioned and so command line help should also mention it.

glassez commented on src/app/options.cpp:
```diff
@@ -244,6 +244,7 @@ namespace
     constexpr const BoolOption NO_SPLASH_OPTION = {"no-splash"};
     constexpr const IntOption WEBUI_PORT_OPTION = {"webui-port"};
     constexpr const StringOption PROFILE_OPTION = {"profile"};
+    constexpr const BoolOption RELATIVE_FASTRESUME = {"relative-fastresume"};
```
IMO, "relative fastresume" is wrong wording.

glassez commented:
I don't like this idea of portable mode. There are too many contradictions. Or it is necessary to apply some restrictions to make it consistent (for example, allow to store portable torrents only in a single directory, specified explicitly or by default in the subdirectory of the profile, etc.).

glassez commented:
Frankly, the use applications of this type in portable mode is generally suspicious for me.

Chocobo1 commented:
>allow to store portable torrents only in a single directory, specified explicitly or by default in the subdirectory of the profile, etc

+1, thought about this. For example a folder called "Download" located at the same folder as qbt.exe or in the Profile folder is good and easy for users to follow.

evsh commented:
>> allow to store portable torrents only in a single directory, specified explicitly or by default in the subdirectory of the profile, etc

> +1, thought about this. For example a folder called "Download" located at the same folder as qbt.exe or in the Profile folder is good and easy for users to follow.

Can not understand your concerns, but want to note that, IMO, users with portable drivers most likely would prefer to keep the qBt in a subfolder, because they use their drives for, well... portable data and qBt may be just a tool, but not a *ding an sich*.

sledgehammer999 commented:
I tried to follow the out-of-code comments and I hope I didn't miss something.
1. There shouldn't be 2 switches for setting the profile dir. Only one. If the user chooses to override the profile path then he should specify the exact path(absolute or relative to current path).
2. An interesting question was "What happens if user downloads something in computer A and moves to continue downloading in computer B"? And something about relative paths. Maybe we should investigate what other clients do(if they have portable mode). Saving as relative path should be done if the files are saved on the same media as the qbt binary. But this is difficult to detect on linux, right?

I haven't looked at the code, so what is the current situation on the above items? Have you come on a solution? Or is this too hard to implement correctly? I wouldn't mind having the option to only set the profile dir without any other assurances.

glassez commented:
@evsh "to store portable torrents only in a single directory" is main idea of my comment. The second thing is "specified explicitly or by default in the subdirectory of the profile" so I think it covers your case.
Also we should disable incomplete folder for portable torrents and probably apply some other restrictions.

glassez commented:
>I wouldn't mind having the option to only set the profile dir without any other assurances.

+1
Since portable mode is more complex task and it requires very serious consideration to do it correctly.

evsh commented:
> 1. There shouldn't be 2 switches for setting the profile dir. Only one. If the user chooses to override the profile path then he should specify the exact path(absolute or relative to current path).

There is only one.

> An interesting question was "What happens if user downloads something in computer A and moves to continue downloading in computer B"?...

On Windows relative paths are written into the .fastresume files if the profile dir and the save path is on the same drive and always on the other platforms. 

> Also we should disable incomplete folder for portable torrents and probably apply some other restrictions.

Why not require it to be inside of the profile dir?

evsh commented:
PR update addressing the raised questions and comments. 

Perhaps we should split it and merge the first five commits and think about the portable mode and fastresume later? I very much want to get configurations support for debugging purposes.

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,399 @@
+    // Base option class. Encapsulates name operations.
+    class Option
+    {
+
```
Remove this blank line.

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,399 @@
+
+        QString fullParameter() const
+        {
+            return QLatin1String("--") + QLatin1String(m_name);
```
You seem to misunderstand the QLatin1String meaning, once you use it for the wrong purpose. Re-read the documentation.

glassez replied:
This applies not only to this line.

evsh replied:
> Re-read the documentation.

Done. Your turn :)

glassez replied:
Ok. Let's go.
>Many of QString's member functions are overloaded to accept const char \* instead of QString. [...] These functions are usually optimized to avoid constructing a QString object for the const char \* data.

>Applications that define QT_NO_CAST_FROM_ASCII don't have access to QString's const char \* API. To provide an efficient way of specifying constant Latin-1 strings, Qt provides the QLatin1String, which is just a very thin wrapper around a const char \*. 

>**Note:** If the function you're calling with a QLatin1String argument isn't actually overloaded to take QLatin1String, the implicit conversion to QString will trigger a memory allocation, which is usually what you want to avoid by using QLatin1String in the first place. In those cases, using QStringLiteral may be the better option.

evsh replied:
Sorry, I've told you already that I did read the documentation. You see, I don't understand what is wrong. Either explain that or give a hint suggesting the right code fragment, please.

glassez replied:
>Sorry, I've told you already that I did read the documentation. You see, I don't understand what is wrong. Either explain that 

Since we don't use QT_NO_CAST_FROM_ASCII using QLatin1String in our code is at least useless (really this leads to unnecessary cost).
>or give a hint suggesting the right code fragment, please

Please. For this particular case:
```c++
const QString DDASH("--");

// ...

QString fullParameter() const
{
    return DDASH + m_name;
}
```

glassez replied:
For other case:
```c++
bool value(const QProcessEnvironment &env) const
{
    QString val = env.value(envVarName());
    // we accept "1" and "true" (upper or lower cased) as boolean 'true' values
     return ((val == "1") || (val.toUpper() == "TRUE"));
 }
```

glassez replied:
Some other case:
```c++
QString envVarName() const
{
    return QString("QBT_") + QString(m_name).toUpper().replace('-', '_');
}
```

evsh replied:
Your code has the following shortcomings:
1. Global `QString` constants lead to `QString` constructor calls at module loading time. Static variables in functions would be a bit better. 
2.  Both local static and global `QString` variables need twice the memory comparing to `char` arrays (permanent usage, is never freed).
3. An expression like `QString("a literal")` requires more computing resources as comparing to `QLatin1String("the same literal"` because UTF-8 to UTF-16 decoder is more complex than ASCII to UTF-16.

And what are the advantages? 

> Since we don't use QT_NO_CAST_FROM_ASCII using QLatin1String in our code is at least useless (really this leads to unnecessary cost).

What costs?

glassez replied:
>Global QString constants lead to QString constructor calls at module loading time. Static variables in functions would be a bit better. 

I don't mean global constant. I mean some constant, for example class-wide.
>Both local static and global QString variables need twice the memory comparing to char arrays (permanent usage, is never freed).

If you feel so sorry for the two bytes, you can do so:
```c++
QString fullParameter() const
{
    return QString("--") + m_name;
}
```

>An expression like QString("a literal") requires more computing resources as comparing to QLatin1String("the same literal") because UTF-8 to UTF-16 decoder is more complex than ASCII to UTF-16.

Agree. But you don't understand the following:
    `QString("str1") + "str2"` = 2 QString ctor + 1 operator+
    `QLatin1String("str1") + QLatin1String("str2")` = 2 QLatin1String ctor + 2 QString ctor + 1 operator+

>What costs?

1. Performance: additional (and redundant) QLatin1String constructor calls
2. Visual: additional (and redundant) QLatin1String in code

evsh replied:
The thing is QLatin1String ctor +  QString ctor is not more work than QString ctor(char literal), because we remove information about encoding. I suggest you read `QLatin1String` class definition, it is quite short. And the general note: it is usually not possible to make something more optimal removing the information, but the inverse case (more optimal with additional information) is more likely. I bet the compiler generates not significantly different code for `QString::fromLatin1("a literal")` and `QString(QLatin1String("a literal"))`.

glassez replied:
Damn. Unfortunately Qt5 documentation is outdated and contains incorrect and conflicting information. It turns out that QString's **const char \*** API is no longer so lightweight and optimized as described.

evsh replied:
I don't know what you are talking about but they say in the documentation that `setCodecForCString()` was removed and all `const char*` strings are assumed to be in UTF-8. Do they hide anything else?

glassez replied:
But the following is wrong:
>Many of QString's member functions are overloaded to accept const char * instead of QString. [...] These functions are usually optimized to avoid constructing a QString object for the const char * data.

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,399 @@
+        {
+        }
+
+        bool operator == (const QString &arg) const
```
The same issue with method names as in another PR.

glassez commented on src/app/options.cpp:
```diff
@@ -0,0 +1,399 @@
+        int value(const QProcessEnvironment &env, int defaultValue) const
+        {
+            QString val = env.value(envVarName());
+            if (val.isEmpty()) return defaultValue;
```
It's better to have blank line after earlier return.

evsh replied:
Done.

glassez commented on src/base/utils/environment.h:
```diff
@@ -0,0 +1,100 @@
+            Downloads
+        };
+
+        class Profile
```
When I suggested Utils::Environment, I proceeded from the sense of the old functions. Environment is something external to the application, but the Profile is not. So you shouldn't put it to Utils::Environment namespace. Let it be in global namespace and its files in `base/` directly.

evsh replied:
While I certainly agree with you that the code does not fall exactly into the "environment" group, it makes an individual unit and I think it would be nice to highlight that by using a proper place in the files hierarchy. Can not find a word right away.

LordNyriox replied:
@evsh, @glassez:  Status?  Location?  ProgramDirectory?

evsh replied:
Sorry, @LordNyriox, those are good names for the entity which is currently named `class Profile`, but we are looking for a name for the namespace (and the corresponding directory), which contains this entity, its various implementations and other things which control how the qBt is localised within a filesystem at run-time.

LordNyriox replied:
@evsh:  Can you try to explain what you mean by "localised within a filesystem at run-time"?  I might be more help if I had a better context.  

Thanks.

evsh replied:
With this PR we are adding options to control were qBt stores its configuration and data files (like the fast resume files or the RSS cache) and we called this entity "Profile". Thank you for the willingness to help.

LordNyriox replied:
@evsh:  ProgramMetadata?

glassez replied:
@LordNyriox Profile is usual name for such a thing. Why do you propose changing it?

LordNyriox replied:
@glassez:  I'm referring to the `Utils:Environment` piece.  If you object to that name, the names I was proposing could each represent the concept equally clearly.

glassez replied:
Things under Utils namespace are helpers, wrappers (and so on) for different purposes. But Profile is more important, so I see it as the application component (albeit small). Therefore, I propose to put it in the global namespace.

evsh replied:
Maybe it will be better to leave implementation classes in namespace Impl?  Perhaps we can create `base/private` or `base/impl` directory and put those files there?

glassez replied:
Agree.
`base/private` and `Private` namespace (as done in BitTorrent).

evsh commented:
PR updated: moved the `Profile` class into the global namespace, implementation classes into namespace `Private`. Moved files to `base/` and `base/private/` respectively.

evsh commented:
@qbittorrent/qbittorrent-frequent-contributors:
I really-really want this to be merged (at least partially). This would make debugging so much simpler! What can I do to make it happen?

glassez commented:
For me, there is no portable mode should be here. I'm really not sure it's implemented correctly now. It requires more thinking about so it's another task.

evsh commented:
@glassez, thank you for the opinion, but this is not an answer to my question. Let me remind the initial motivation: to make development easier it makes sense to load different list of torrents in the debugging mode than that of a regular run. 
Attempts to implement a full-fledged portable mode did not only improved the initial implementation, but also raised requirements quite high. And the latter complicates any efforts to merge this code into the master branch. That is why my questions is: what can be done to merge at least the configurations support?

glassez commented:
@evsh I don't understand what does not suit you in my answer... You say that "portable mode" prevents merging this request:
>Attempts to implement a full-fledged portable mode did not only improved the initial implementation, but also raised requirements quite high. And the latter complicates any efforts to merge this code into the master branch.

So you just need to remove it from here and leave only "configuration" part.

Owyn commented:
> command line

so no support for file association? 

I mean when you associate .torrent file with qbittorrent - it won't use any command lines switches, would it?

evsh commented:
@sledgehammer999: would you accept this PR without the very last commit? Or, better, with it, please :)

evsh commented:
@sledgehammer999 ping?

evsh commented:
@sledgehammer999: ping? Don't you agree that profiles would make debugging so much simpler?

evsh commented:
@sledgehammer999 ping?

evsh commented:
@sledgehammer999 ping again?

sledgehammer999 commented:
@evsh can you rebase. There are merge conflicts.

evsh commented:
Rebased

sledgehammer999 commented:
@evsh I am actively reviewing this. If you have time, rebase. I think the new rand() PR was the cause.

evsh commented:
@sledgehammer999 rebased.

sledgehammer999 commented:
Please do a final rebase. The qt4 drop messed up 2 files.

sledgehammer999 commented:
@evsh any news?

evsh commented:
Oh, missed the previous message, sorry. Will rebase right away.

evsh commented:
Rebased

evsh commented:
Corrected commit 11b93d6 (some changes from commit fb79260, the next one) were mistakenly put into it (during a rebase, probably).

evsh commented:
And did it one more time.

evsh commented:
@sledgehammer999: are there any news or problems?

glassez commented:
Apparently, the main problem is the lack of @sledgehammer999 free time. IMO, this is the "bottleneck" of the project.

sledgehammer999 commented on src/app/application.cpp:
```diff
@@ -93,25 +94,40 @@ namespace
     SettingsStorage::initInstance();
     Preferences::initInstance();
 
+    if (m_commandLineArgs.webUiPort > 0) {
```
you already check this in `parseCommandLine()`

evsh replied:
This test meant to show whether the port number is set by user, or has its default value of -1. Probably should change condition to `!= -1`?

sledgehammer999 replied:
Oops. You're correct. I forgot that this might be unset. In this case this check can stay as it is.

sledgehammer999 commented on src/app/options.cpp:
```diff
@@ -0,0 +1,192 @@
+            else if (arg == QLatin1String("--profile")) {
+                QStringList parts = arg.split(QLatin1Char('='));
+                if (parts.size() == 2)
+                    result.profileDir = parts.last();
```
what happens if it is malformed? No 2 parts?

sledgehammer999 commented on src/app/options.cpp:
```diff
@@ -0,0 +1,192 @@
+#endif
+            else if (arg == QLatin1String("--profile")) {
+                QStringList parts = arg.split(QLatin1Char('='));
+                if (parts.size() == 2)
```
what happens if it is malformed? No 2 parts?

sledgehammer999 replied:
I see it will stay empty so it will fallback to using the DefaultProfile.

sledgehammer999 commented on src/app/options.cpp:
```diff
@@ -0,0 +1,192 @@
+            }
+            else if (arg == QLatin1String("--configuration")) {
+                QStringList parts = arg.split(QLatin1Char('='));
+                if (parts.size() == 2)
```
what happens if it is malformed? No 2 parts?

sledgehammer999 replied:
OK, defaults are used if empty

sledgehammer999 commented on src/app/options.h:
```diff
@@ -1,5 +1,7 @@
 /*
- * Bittorrent Client using Qt4 and libtorrent.
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2016  Eugene Shalygin
```
I don't know if this intended or not. This is just a reminder. You might want to leave an email beside your name.
If yes, add it on every other place you put your name.

evsh replied:
At the time of writing I did not want to leave e-mail there.  Probably I'll insert the e-mail during the next round of changes.

sledgehammer999 replied:
Sure. Whatever you want. It was just just a reminder.
(I am continuing with the review)

sledgehammer999 commented:
Only the last commit remains. Will be done tomorrow when I have a non-sleepy and clear mind. I expect to find minimal problems. So @evsh if in the meantime you want to address things from the 1st commit feel free to push changes to it.

evsh commented on src/base/private/profile_p.cpp:
```diff
@@ -0,0 +1,208 @@
+SettingsPtr Private::CustomProfile::applicationSettings(const QString &name) const
+{
+    // here we force QSettings::IniFormat format always because we need it to be portable across platforms
+    const QString settingsFileName {QDir(configLocation()).absoluteFilePath(name + QLatin1String(".conf"))};
```
Disagree with you: current files have '.conf' extension, and if we use '.ini' in the custom profiles, we make a custom profile dir incompatible with the current one, while otherwise profiles dirs can be copied one into another.

sledgehammer999 replied:
>current files have '.conf' extension

Partially true. The old code has an ifdef when creating the QSettings object. In linux it uses NativeFormat, which are ini files with `.conf` extension. On Windows+macOS it uses IniFormat which forces `.ini` as extension.

sledgehammer999 commented on src/base/profile.cpp:
```diff
@@ -47,14 +48,20 @@ Profile::Profile(Private::Profile *impl)
+
+    QScopedPointer<Private::PathConverter> converter(convertPathsToProfileRelative
+                                                     ? static_cast<Private::PathConverter *>(new Private::NoConvertConverter())
+                                                     : static_cast<Private::PathConverter *>(new Private::Converter(profile->baseDirectory())));
```
Shouldn't the arguments be reversed here? (I am talking about the bool check)

evsh commented on src/app/application.cpp:
```diff
@@ -660,4 +661,7 @@ void Application::validateCommandLineParameters()
         throw CommandLineParameterError(tr("Portable mode and explicit profile directory options are mutually exclusive"));
+
+    if (m_commandLineArgs.portableMode && m_commandLineArgs.relativeFastresumePaths)
+        throw CommandLineParameterError(tr("Portable mode and explicit relative fastresume mode options are mutually exclusive"));
```
> IMO, we should silently ignore...

Wouldn't that create an impression that `--relative-fastresume` is used somehow? Let's print a warning instead?

sledgehammer999 replied:
Well, if they read the documenation they should have that impression. But I am ok with a warning.

evsh commented:
> Partially true. The old code has an ifdef when creating the QSettings object. In linux it uses NativeFormat, which are ini files with .conf extension. On Windows+macOS it uses IniFormat which forces .ini as extension.

Shall I then add the same ifdef here?

sledgehammer999 commented:
>Shall I then add the same ifdef here?

If you do then it won't be cross-platform as you want. (linux files won't load on Windows+macOS and vice versa)

evsh commented:
> it won't be cross-platform as you want.

I, frankly, do not remember my intentions regarding cross-platform profiles... But if it seems possible, why not? Then we need to leave .conf extension, right?

sledgehammer999 commented:
>I, frankly, do not remember my intentions regarding cross-platform profiles... But if it seems possible, why not? Then we need to leave .conf extension, right?

I don't understand why you are so against having `.ini` extension in profile case.
As for the use case, IMO it makes sense for someone to be able to transfer profiles between PCs/operating systems. Especially in the case of `--portable`.

evsh commented:
> I don't understand why you are so against having .ini extension in profile case.

No, I'm for the same extension for settings files whether they are portable profile or in the $HOME/.config. IMO, copying profiles between different OSes is less frequent operation than copying them into each other (making a portable profile default or vise versa) within the same OS.

sledgehammer999 commented:
>No, I'm for the same extension for settings files whether they are portable profile or in the $HOME/.config.

IIRC, the default profile will continue to use the old ifdefs anyway. Question is what we do about the new cases?

>IMO, copying profiles between different OSes is less frequent operation than copying them into each other (making a portable profile default or vise versa) within the same OS.

I agree. But why not do portability across OSes possible too? Is it because the user can't copy those files into the default path in case of Linux? (assuming he wants to make some profile his main profile).
If that's the reason, then switch to idefs for this case too instead of hardcoding the extension.

evsh commented:
> But why not do portability across OSes possible too? Is it because the user can't copy those files into the default path in case of Linux? 

Do you mean copying `qBittorrent.ini` into `qBittorrent.conf`, which is of course possible, because both are in the same format. But is this convenient? For a user in this situation it would be convenient, perhaps, if app takes whichever file it finds, .ini or .conf. But what to do if there are two? Maybe indeed take this way and signal error if there are two configuration files in a profile?

sledgehammer999 commented:
Let me rephrase:
If we want seamless portability between OSes we need to use the same file extension for all OSes. I prefer `.ini` since this is what IniFormat uses.
The caveat of this is that the Linux users cannot interchange between their default and custom profiles manually(making a custom profile default) by just dropping the files to the correct path.
So we are left with 2 options:
1. Go ahead and use `.ini` for non-default profiles and let the linux users rename their extensions manually when they wish to interchange manually profiles.
2. Don't actually implement seamless OSes portability and use ifdefs to ensure portability only between the same type of OS. Users can still rename the necessary files manually.

I don't have any preference between those 2 options. So I leave the choice upon you. Make the choice, do any other needed changes, push and comment so I can merge this.

glassez commented:
You guys are stalled in one place. If you want to make it portable, it must have the same extension regardless of platform. It remains only to choose an .ini or .conf (it does not matter, purely subjective question).
>The caveat of this is that the Linux users cannot interchange between their default and custom profiles manually(making a custom profile default) by just dropping the files to the correct path.

Why can't linux users use the same extension for default and custom profiles? Let's use .ini everywhere! Or let's use .conf. What's the real problem here?

sledgehammer999 commented:
>Why can't linux users use the same extension for default and custom profiles? Let's use .ini everywhere! Or let's use .conf. What's the real problem here?

The problem comes from the current way(before this PR) of saving. Linux uses `NativeFormat` which uses the .conf extension while the Windows+macOS use `IniFormat` which uses the .ini extension.

evsh commented:
Would suggest .conf everywhere and migration to JSON later.

sledgehammer999 commented:
I predict this will start going in circles and result in more delays and possibly new code to write.
So it is better(for the time being) to write `Private::CustomProfile::applicationSettings()` as:
```c++
SettingsPtr Private::CustomProfile::applicationSettings(const QString &name) const
{
    // here we force QSettings::IniFormat format always because we need it to be portable across platforms
#if defined(Q_OS_WIN) || defined(Q_OS_MAC)
    const QString settingsFileName {QDir(configLocation()).absoluteFilePath(name + QLatin1String(".ini"))};
#else
    const QString settingsFileName {QDir(configLocation()).absoluteFilePath(name + QLatin1String(".conf"))};
#endif

And we can see in the future if we migrate to something universal for all configs.
    return SettingsPtr(new QSettings(settingsFileName, QSettings::IniFormat));
}
```

evsh commented:
OK

evsh commented:
All the comments were addressed, I believe.

sledgehammer999 commented:
I believe this is ready. Thank you all for the code, reviews, suggestions etc.

evsh commented:
Oh, yes! Thanks to all of you! 

Was never happier with a single merged PR before, despite I'm pretty sure that this one will require a lot of post-merge care :)
